ARC: interrupt management fix for SMP case

evgeniy-paltsev opened the pull request:
**NOTE**: one part of the described issues is already fixed & merged as a separate PR #52387. The priority issue is still a thing and this PR can be used as a reference about how we can workaround this.
--------------------------


## issue description

In case of SMP system interrupt infrastructure implementation for ARC has serious design flaw.

We have per-core interrupt controllers private for each core. They are used to control interrupts - enable/disable, set priority. The IDU (interrupt distribution unit) is setup to distribute interrupts to different cores in the round-robin manner.

The irq management functions (`irq_enable`, `irq_disable`, `irq_is_enabled`) setup/check interrupt controller state only for the core where they are called. This cause inconsistency - we may have interrupt enabled on one core and disabled it on another. Such approach makes common interrupts unusable in case of SMP - there is no correct way to enable/disable interrupt line, we will lose some interrupts if we simple use `irq_enable` or we will get unexpected interrupts if we use `irq_disable`. See #31298 for example.

## reworked implementation description

We manage interrupts (enable/disable) with IDU instead of per-core interrupt controller for the common interrupts. The per-core interrupts enabling/disabling is still managed by per-core interrupt controller.

The interrupt priority can't be managed by IDU so we still control it with per-core interrupt controller. We setup interrupt priority at init stage. For that we collect interrupt priority of static interrupts in compile time. We don't support dynamic interrupts for the common interrupts for SMP systems for now.

We keep interrupt handling on all CPUs, and we do initial interrupt controller setup on all CPUs and enable round-robin interrupt distribution mode in IDU. However we move IDU setup from SoC code to generic ARC code.

NOTE: this patch doesn't include arc interrupts initialization optimization & cleanup (such as removing of excessive locking in core / IDU interrupts setup functions and implementation of proper *_nolock functions analogs). This will be done in separate PR.

Fixes: #31415
Fixes: #31298

evgeniy-paltsev committed:
ARC: control shared (common) interrupts via IDU

Control shared interrupts enabling/disabling via IDU.
With that we can easily enable and disable them for all cores
in one place.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: SoC: drop per-SoC IDU setup

We don't need to setup IDU at SoC level as we do it in
generic way now.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
tests: ARC: add basic test for interrupt controller functionality

Test isr priority section, initial interrupt states, static
IRQ_CONNECT functionality.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

ruuddw commented on arch/arc/include/arc-connect-priv.h:
```diff
@@ -0,0 +1,14 @@
+/*
```
filename with dash instead of underscore?

evgeniy-paltsev replied:
this file doesn't exist in updated version.

ruuddw commented on include/arch/arc/v2/arcv2_irq_unit.h:
```diff
@@ -119,19 +119,13 @@ bool z_arc_v2_irq_unit_int_enabled(int irq)
 	unsigned int key = arch_irq_lock();
 
 	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
-#if defined(CONFIG_ARC_SECURE_FIRMWARE)
```
Why is this no longer needed? Is this _raw only used in case of ARC_CONNECT/SMP so Secure Mode is not relevant?

evgeniy-paltsev replied:
Secure mode is still relevant, however this functionality has been split into two parts 
- the setting of raw value in hardware. It is done by `z_arc_v2_irq_unit_prio_set_raw`
- converting interrupt priority value into the raw value which should be written to HW (with taking care of secure bit and priority split for secure / normal). It is done by `ARC_IRQ_PRIO_TO_HW` macro.

PS: and yes, now this function is used only in interrupt controller init code.

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -40,6 +40,21 @@ extern void z_irq_priority_set(unsigned int irq, unsigned int prio,
 extern void z_irq_spurious(const void *unused);
 
+
+struct arc_isr_sratic_data {
```
typo sratic -> static

evgeniy-paltsev replied:
fixed (we don't have this structure in updated version)

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -84,7 +101,9 @@ extern void z_irq_spurious(const void *unused);
 	"irq priority cannot be set to 0 when CONFIG_ARC_FIRQ_STACK" \
 	"is not configured or CONFIG_ARC_FIRQ_STACK " \
 	"and CONFIG_ARC_STACK_CHECKING are configured together"); \
-	z_irq_priority_set(irq_p, priority_p, flags_p); \
```
Some comments why priority cannot be set here and is moved would be good (I don't really understand why it cannot be left here). Also, above comments might not be applicable at this location, when initialization is moved (not sure I understand the comment, or why it is a comment and not an assert).

evgeniy-paltsev replied:
Updated the comment for ARCH_IRQ_CONNECT, thanks!

ruuddw commented on tests/arch/arc/interrupt-controller/src/interrupt-controller.c:
```diff
@@ -0,0 +1,188 @@
+
+
+static void intc_irq_check_masking_dynamic_single(uint32_t irq)
+{
```
Is testing via the get/set functions enough, or should there be some real interrupt triggering test to see if the get/set are implemented correctly?

ruuddw commented on tests/arch/arc/interrupt-controller/src/interrupt-controller.c:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+static void dummu_irq_handler0(const void *unused) {}
```
typo, dummu -> dummy

evgeniy-paltsev replied:
fixed

ruuddw commented:
General principle is fine to fix the SMP interrupt management (for dynamic interrupts, some IPI would be required to set priorities, but I'm fine with this static solution).
See few additional comments inline.

evgeniy-paltsev committed:
gen_isr_tables: introduce interrupt priority table generation

Introduce CONFIG_GEN_IRQ_PRIORITY_TABLE option which controls
whether a platform using gen_isr_tables needs a interrupt
priorities table created. This table is an array with priorities
of static interrupts connected via IRQ_CONNECT macro. The array has
same size as IRQ_VECTOR_TABLE and it is sorted by interrupt line
number.

The option is disabled by default, so it doesn't interfere with
the architectures using gen_isr_tables unless explicitly enabled.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: setup interrupt priority of static irqs at init time

Save interrupt priority at compile time with
GEN_IRQ_PRIORITY_TABLE mechanism so we can init it for all
interrupts at core interrupt controller init phase.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
test: gen_isr_table: test interrupt priority table generation

Add check of interrupt priority table generation in case of
this feature (CONFIG GEN_IRQ_PRIORITY_TABLE) is enabled.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
tests: gen_isr_table: disable dynamic IRQs in case of ARC_CONNECT

Disable dynamic interrupts test for ARC in case of ARConnect
presence (CONFIG_ARC_CONNECT) as we don't support dynamic
interrupts for shared interrupts (interrupts routed via IDU
in ARConnect) for ARC.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

abrodkin commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -17,6 +17,11 @@ extern uint32_t _irq_vector_table[];
 #define HAS_DIRECT_IRQS
 #endif
 
+/* ARC don't support dynamic interrupts on the systems with ARConnect */
```
I'd add more justification here. As you put it in the PR message for example:
> We don't support dynamic interrupts for the common interrupts for SMP systems for now.
Moreover maybe add even more details, why's that.

evgeniy-paltsev replied:
Done.

abrodkin reviewed (changes requested):
Even though I do like that change and obviously we badly need it for ARC SMP systems I think it would be nice to add some introduction of IRQ controllers hierarchy in ARC SMP systems. Preferably with some diagram so that reader understands what do we do here and why. Something similar to what we have here: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arc/plat-hsdk/platform.c but maybe with 2 ARC cores depictured.

evgeniy-paltsev commented on include/arch/arc/v2/irq.h:
```diff
@@ -40,18 +40,20 @@ extern void z_irq_priority_set(unsigned int irq, unsigned int prio,
 #define ARCH_IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p) \
 { \
-	Z_ISR_DECLARE(irq_p, 0, isr_p, isr_param_p); \
-	z_irq_priority_set(irq_p, priority_p, flags_p); \
```
Hi Ruud,

please don't forget that we are moving info from one place to another :)

I've done some measurements:

---
If we actually use interrupt lines (if we connect handlers to them) there is even a image size decrease:

Setup with IRQ used, hsdk board (CONFIG_NUM_IRQS=88), without changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32728 B      1280 MB      0.00%
```

Setup with IRQ used, hsdk board (CONFIG_NUM_IRQS=88), with changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32124 B      1280 MB      0.00%
```

Setup with IRQ used, nsim_hs board (CONFIG_NUM_IRQS=30), without changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9896 B         1 MB      0.94%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ used, nsim_hs board (CONFIG_NUM_IRQS=30), with changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9780 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

----

However if we have a setup with a lot of interrupts which are **not used**, there is a small penalty:

Setup with IRQ unused, nsim_hs board (CONFIG_NUM_IRQS=30), without changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9768 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ unused, nsim_hs board (CONFIG_NUM_IRQS=30), with changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9776 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ unused, hsdk board (CONFIG_NUM_IRQS=88), without changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32088 B      1280 MB      0.00%
```


Setup with IRQ unused, hsdk board (CONFIG_NUM_IRQS=88), with changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32120 B      1280 MB      0.00%
```

The good news is that we can strip interrupt table (and interrupt priority table) if we don't use some of the interrupts.

There are another possible optimizations:
- we can disable interrupt priority table generation if we have only one interrupt priority level.
- we can introduce interrupt priority table packing (i.e on ARC we have 16 interrupt priorities levels, so is enough to have 4bits to store them)

However, I guess it would be better to have these optimizations in a separate PR.

ruuddw replied:
Great to see the numbers, indeed the trade-off is between constant data and code. The increase without (many) ISRs installed is very small, let's avoid more #ifdef's. Other optimizations would (like you suggest) have more impact in that case.

abrodkin commented on drivers/interrupt_controller/intc_arcv2_irq_unit.c:
```diff
@@ -20,44 +20,112 @@
+
+	for (uint32_t i = 0; i < (CONFIG_NUM_IRQS - ARC_CONNECT_IDU_IRQ_START); i++) {
+		/*
+		 * TODO: don't use z_arc_connect_idu* finctions to avoid
```
finctions -> fUnctions

evgeniy-paltsev replied:
fixed

abrodkin commented on drivers/interrupt_controller/intc_arcv2_irq_unit.c:
```diff
@@ -20,44 +20,112 @@
- * selected IRQ's registers.
- *
- * @return 0 for success
+ * NOTE: core interrupts are still locked at this point, so there is no need to protect the window
```
"interrupts are still locked"? Maybe "masked" or even disabled? And do you mean they are disabled individually or global `STATUS32.IE` flag is not set yet?
For the record, IIRC all IRQs are enabled by default in the private ARC INTC.

evgeniy-paltsev replied:
The interrupts are globally disabled (STATUS32.IE = 0).
I've updated the comment.

abrodkin reviewed (dismissed):
With minor comments on comments LGTM.

ruuddw commented:
@evgeniy-paltsev ok to remove the draft label? I think this is ready for broader review and merging now.

mbolivar-nordic commented:
@evgeniy-paltsev any plans to update this?
